src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Setup/SetupComponentProvider.cs

```diff
@@ -0,0 +1,192 @@
+/// to <see cref="ISetupComponentRegistrationService"/> (global scope), which aggregates across
+/// all projects and notifies the user to install missing components via in-product acquisition.
+/// </summary>
+/// <remarks>
+/// Reported requirements are:
+/// <list type="bullet">
+///   <item><see cref="SuggestedWorkload"/> items from the <c>CollectSuggestedWorkloads</c> target (yay!).</item>
+///   <item>Specific workloads based on project capabilities and hard-coded knowledge about project types and .NET features (boo!).</item>
+///   <item>The .NET runtime version (for .NET Core project configurations only).</item>
+/// </list>
+/// Components are gathered from all active configured projects within the project.
+/// </remarks>
+[Export(ExportContractNames.Scopes.UnconfiguredProject, typeof(IProjectDynamicLoadComponent))]
+[AppliesTo(ProjectCapability.DotNet)]
+internal sealed class SetupComponentProvider : OnceInitializedOnceDisposedAsync, IProjectDynamicLoadComponent
+{
+    private static readonly IImmutableSet<string> s_evaluationRuleNames = ImmutableStringHashSet.EmptyRuleNames.Add(ConfigurationGeneral.SchemaName);
+    private static readonly IImmutableSet<string> s_buildRuleNames = ImmutableStringHashSet.EmptyRuleNames.Add(SuggestedWorkload.SchemaName);
+
+    private readonly UnconfiguredProject _unconfiguredProject;
+    private readonly ISafeProjectGuidService _safeProjectGuidService;
+    private readonly ISetupComponentRegistrationService _setupComponentRegistrationService;
+    private readonly IActiveConfiguredProjectSubscriptionService _activeConfiguredProjectSubscriptionService;
+    private readonly IActiveConfigurationGroupService _activeConfigurationGroupService;
+    private readonly IProjectFaultHandlerService _projectFaultHandlerService;
+    private readonly DisposableBag _disposables = new();
+
+    private Guid _projectGuid;
+
+    [ImportingConstructor]
+    public SetupComponentProvider(
+        UnconfiguredProject project,
+        ISafeProjectGuidService safeProjectGuidService,
+        ISetupComponentRegistrationService setupComponentRegistrationService,
+        IActiveConfiguredProjectSubscriptionService activeConfiguredProjectSubscriptionService,
+        IActiveConfigurationGroupService activeConfigurationGroupService,
+        IProjectFaultHandlerService projectFaultHandlerService,
+        IProjectThreadingService threadingService)
+        : base(threadingService.JoinableTaskContext)
+    {
+        _unconfiguredProject = project;
+        _safeProjectGuidService = safeProjectGuidService;
+        _setupComponentRegistrationService = setupComponentRegistrationService;
+        _activeConfiguredProjectSubscriptionService = activeConfiguredProjectSubscriptionService;
+        _activeConfigurationGroupService = activeConfigurationGroupService;
+        _projectFaultHandlerService = projectFaultHandlerService;
+    }
+
+    public Task LoadAsync()
```

Comment:
When does this method get called? What is the capabilities condition that it would be looking for?

Reply:
This is an implementation of `IProjectDynamicLoadComponent.LoadAsync`. Exports of `IProjectDynamicLoadComponent` are loaded/unloaded in response to project capabilities coming/going. In this case, the `.NET` capability is static, so won't "go", and we no-op in the `UnloadAsync` method. It would be called when the project unloads, but then this class is disposed at that time anyway.

The use of `IProjectDynamicLoadComponent` here replaces the use of `IPackageService` we had previously, which initializes components when the package loads. Package load can happen quite "late". For example, you can load a solution and be coding/compiling/etc long before our package ever loads. We want to notify straight away if components are missing, so use `IProjectDynamicLoadComponent`. That's one of the improvements in this PR.